gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsBatch.java

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud.dns;
+
+import com.google.gcloud.Page;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * A batch of operations to be submitted to Google Cloud DNS using a single HTTP request.
+ */
+public class DnsBatch {
+
+  private Map<Request, Callback> requests = new LinkedHashMap<>();
+  private Dns dns;
+
+  /**
+   * An interface for the callback which will be invoked when the operation has been executed. The
+   * parameter {@code <T>} represents the type of the result of the operation and thus depends on
+   * the {@link DnsBatch.Request} that this call back belongs to and it should be as follows:
+   *
+   * <ul>
+   *   <li>{@link Zone} for creating and getting a zone</li>
+   *   <li>{@link Boolean} for deleting a zone</li>
+   *   <li>{@link ChangeRequest} for creating and getting a change request</li>
+   *   <li>{@link ProjectInfo} for getting a project</li>
+   *   <li>{@code Page<Zone>} for listing zones</li>
+   *   <li>{@code Page<DnsRecord>} for listing {@link DnsRecord}s inside a zone</li>
+   *   <li>{@code Page<ChangeRequest>} for listing {@link ChangeRequest}s for a zone</li>
+   * </ul>
+   */
+  public interface Callback<T> {
+    /**
+     * A method which will be called if the {@link DnsBatch.Request} succeeds. See the {@link
+     * Callback} documentation for details on type {@code T}.
+     *
+     * @param output the result of the operation
+     * @param request the request which succeeded
+     */
+    void success(T output, DnsBatch.Request request);
+
+    /**
+     * A method which will be called if the {@link DnsBatch.Request} fails.
+     *
+     * @param ex the error
+     * @param request the request which failed
+     */
+    void error(DnsException ex, DnsBatch.Request request);
+  }
+
+  /**
+   * An operation to be submitted to Google Cloud DNS within this batch. Only an subset of the class
+   * attributes appropriate for the represented operation is initialized. Refer to the class method
+   * and attribute documentation for the specific fields.
+   */
+  public static class Request {
```

Comment:
> I did consider it. Having an AbstractRequest and inherit from it 8 times seemed kind of overkill for the benefit, especially from the maintenance point of view, so I chose this solution. I would like to hear other opinions

I don't feel strongly about this as users do not create request objects directly so at least the confusion at creation time is avoided. However, even with the nice javadoc you added it might be confusing to have many `null` fields in `Request`. Let's wait for more opinion.

> Made serializable for now.

Here is my point on this being Serializable. I thing there's a benefit for this class but I also see a benefit in `DnsBatch` being [Restorable](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-core/src/main/java/com/google/gcloud/Restorable.java#L39). However this should not come at the price of making `Callback` serializable (which is too restrictive for users). Thoughts? /cc @aozarov @ajkannan 

> With hashCode and equals, I did not override them on purpose.

This is a good point. And I guess this means that operation in a batch are executed in order, right? If we keep this `Serializable` we must have proper `hashCode` and `equals` methods. Maybe wait for the discussion on `Serializable`/`Restorable` to settle but I would consider a list of pairs (we are not using the "Hash" part of LinkedHashMap now, right?). Also using two separate lists would do no harm.